In [54]:
# Imports
import math

# concat function
def concat(a, b):
    return int(f"{a}{b}")

In [34]:
# Values p and q
p = int(11)
q = int(13)

In [35]:
# N value
n = p * q
print("n =", n)

n = 143


In [36]:
# Phi value
phi = (p - 1) * (q - 1)
print("phi =", phi)

phi = 120


In [65]:
# Function for finding e
def find_e(phi):
    e_values = []
    for i in range(2, phi):
        if math.gcd(i, phi) == 1:
            e_values.append(i)
    return e_values
print ("e values =", find_e(phi))

e values = [7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 49, 53, 59, 61, 67, 71, 73, 77, 79, 83, 89, 91, 97, 101, 103, 107, 109, 113, 119]


In [66]:
# Choose e from the list of e values
e = 7
print("e =", e)    

e = 7


In [63]:
# Function for finding d
def find_d(e, phi):
    d = 1
    while ((e * d) % phi != 1):
        d += 1
    return d
print("d =", find_d(e, phi))

d = 103


##### Math for finding $d$ 
$ d \equiv e^{-1} (mod \phi(n)) $

Eukleides Algorithm  (example for $e = 7$ and $\phi(n) = 120$)
  
| a | b | $a_{i}$ / $b_{i}$ | $a mod b $ | $a mod b = a_{i} - b_{i}$ |
| :-: | :-: | :-: | :-: | :-: |
| $\phi(n) = 120$ | $e = 7$ | $17$ | $1$ | $1 = 1\cdot 120 - 17\cdot 7$ |
| $7$ | $1$ | $7$ | $0$ |  |

$gcd(120, 7) = 1 = 1 \cdot 120 - 17 \cdot 7$  
So $d \equiv x (mod \phi(n)) = x (mod (p-1)\cdot(q-1)) = -17 (mod 120) = 103$  
Negative modulo calculation: $-17 (mod 120)  = 120 - 17 (mod 120) = 120 - 17 = 103$

In [39]:
# Function of finding e if d is known
def find_e_d(d, phi):
    e = 1
    while ((e * d) % phi != 1):
        e += 1
    return e
print("e =", find_e_d(103, phi))

e = 7


##### Math for finding $e$ 
$ d \cdot e \equiv 1 (mod \phi(n)) $

In [64]:
def encrypt_number(list, e, n):
    # Encrypt the number
    number = list[0]
    for i in range(1, len(list)):
        number = concat(number, list[i])   
    encrypted_number = pow(number, e, n)
    # Return the encrypted number
    return encrypted_number

def encrypt(message, e, n, block_size):
    # Convert the message to a list of numbers
    message = list(map(int, list(message)))
    # Create an empty list for the encrypted message
    encrypted_message = []
    # Loop through the message
    for i in range(0, len(message), block_size):
        # Get the block
        block = message[i:i + block_size]
        # Encrypt the block
        encrypted_block = encrypt_number(block, e, n)
        # Add the encrypted block to the encrypted message
        encrypted_message.append(encrypted_block)
    # Return the encrypted message
    return encrypted_message

message = "6425"
block_size = 2    
encrypted_message = encrypt(message, e, n, block_size)
print("encrypted message =", encrypted_message)

encrypted message = [103, 64]


Encrypt equation:  $ c \equiv m^{e} (mod n) $  
*Example for message 6425, block = 2, $e = 7$ and $n = 143$*  
$64^{7} mod 143 = 103$  
$25^{7} mod 143 = 64$    
So the encrypted message is $c = 103$  $64$

In [61]:
# decrypt_block function
def decrypt_block(encrypted_block, d, n):
    # Create an empty list for the decrypted block
    decrypted_block = []
    # Loop through the encrypted block
    for number in encrypted_block:
        # Decrypt the number
        decrypted_number = pow(number, d, n)
        # Add the decrypted number to the decrypted block
        decrypted_block.append(decrypted_number)
    # Return the decrypted block
    return decrypted_block

d = find_d(e, phi)
decrypted_message  = decrypt_block(encrypted_message, d, n)
print("decrypted message =", decrypted_message)    

decrypted message = [64, 25]


Decrypt equation:  $ m \equiv c^{d} (mod n) $  
*Example for message 10364, block = 1 (since odd number of message characters), $d = 103$ and $n = 143$*  
$103^{103} mod 143 = 64$  
$64^{103} mod 143 = 25$    
So the decrypted message is $m = 64$  $25$